src/viam/sdk/common/proto_t.hpp

```diff
@@ -0,0 +1,293 @@
+
+    // Obtain integer constant representing the stored data type.
+    int kind() const {
+        return vtable_.kind();
+    }
+
+    // Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoT&);
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoT&);
+
+   private:
+    struct vtable_t {
+        void (*dtor)(void*) noexcept;
+        void (*copy)(void const*, void*);
+        void (*move)(void*, void*) noexcept;
+        void (*to_proto_value)(void const*, google::protobuf::Value*);
+        int (*kind)() noexcept;
+        bool (*equal_to)(void const*, void const*, const vtable_t&);
+    };
+
+    template <typename T>
+    struct model_t {
+        model_t(T t) : data(std::move(t)) {}
+
+        static void dtor(void* self) noexcept {
+            static_cast<model_t*>(self)->~model_t();
+        }
+
+        static void copy(void const* self, void* dest) {
+            new (dest) model_t(*static_cast<model_t const*>(self));
+        }
+
+        static void move(void* self, void* dest) noexcept {
+            new (dest) model_t(std::move(*static_cast<model_t*>(self)));
+        }
+
+        static void to_proto_value(void const* self, google::protobuf::Value* v);
+
+        static int kind() noexcept;
+
+        static bool equal_to(void const* self, void const* other, const vtable_t& other_vtable) {
+            if (model_t::kind() != other_vtable.kind()) {
+                return false;
+            }
+
+            return *static_cast<T const*>(self) == *static_cast<T const*>(other);
+        }
+
+        static constexpr vtable_t vtable{dtor, copy, move, to_proto_value, kind, equal_to};
+        T data;
+    };
+
+    struct storage_t {
+        static constexpr std::size_t small_size =
+            sizeof(std::unordered_map<std::string, std::string>);
+        template <typename T>
+        storage_t(T t) {
+            static_assert(sizeof(model_t<T>) <= small_size, "Too big!");
+            new (&buf_) model_t<T>(std::move(t));
+        }
+
+        storage_t(const storage_t&) = delete;
+        storage_t(storage_t&&) = delete;
+        storage_t& operator=(const storage_t&) = delete;
+        storage_t& operator=(storage_t&&) = delete;
+
+        storage_t(const storage_t& other, const vtable_t& vtable) {
+            vtable.copy(other.get(), this->get());
+        }
+
+        storage_t(storage_t&& other, const vtable_t& vtable) {
+            vtable.move(other.get(), this->get());
+        }
+
+        void swap(const vtable_t& this_vtable, storage_t& other, const vtable_t& other_vtable) {
+            if (this == &other) {
+                return;
+            }
+
+            unsigned char tmp[small_size];
```

Comment:
I think you're right, was confusing myself a bit reading that `aligned_storage` has recently been deprecated, but realized that they mean the issue is people writing `aligned_storage<N> my_storage` rather than `aligned_storage_t<N> my_storage` etc